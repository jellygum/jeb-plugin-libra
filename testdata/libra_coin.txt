CompiledModule: {
Module Handles: [
	0x0.LibraCoin,]
Struct Handles: [
	resource T@0x0.LibraCoin,
	resource MintCapability@0x0.LibraCoin,]
Function Handles: [
	0x0.LibraCoin.borrow_sender_mint_capability(): (&resource MintCapability@0x0.LibraCoin),
	0x0.LibraCoin.mint(Integer, &resource MintCapability@0x0.LibraCoin): (resource T@0x0.LibraCoin),
	0x0.LibraCoin.grant_mint_capability(): (),
	0x0.LibraCoin.zero(): (resource T@0x0.LibraCoin),
	0x0.LibraCoin.value(&resource T@0x0.LibraCoin): (Integer),
	0x0.LibraCoin.split(resource T@0x0.LibraCoin, Integer): (resource T@0x0.LibraCoin, resource T@0x0.LibraCoin),
	0x0.LibraCoin.withdraw(&mut resource T@0x0.LibraCoin, Integer): (resource T@0x0.LibraCoin),
	0x0.LibraCoin.join(resource T@0x0.LibraCoin, resource T@0x0.LibraCoin): (resource T@0x0.LibraCoin),
	0x0.LibraCoin.deposit(&mut resource T@0x0.LibraCoin, resource T@0x0.LibraCoin): (),
	0x0.LibraCoin.destroy_zero(resource T@0x0.LibraCoin): (),
	0x0.LibraCoin.TODO_REMOVE_burn_gas_fee(resource T@0x0.LibraCoin): (),]
Struct Definitions: [
	{resource T@0x0.LibraCoin
		resource T@0x0.LibraCoin.value: Integer},
	{resource MintCapability@0x0.LibraCoin},]
Field Definitions: [
	resource T@0x0.LibraCoin.value: Integer,]
Function Definitions: [
	public 0x0.LibraCoin.borrow_sender_mint_capability(): (&resource MintCapability@0x0.LibraCoin)
		locals(0): Address, &mut resource MintCapability@0x0.LibraCoin, &resource MintCapability@0x0.LibraCoin,
		GetTxnSenderAddress
		StLoc(0)
		MoveLoc(0)
		BorrowGlobal(1)
		StLoc(1)
		MoveLoc(1)
		FreezeRef
		StLoc(2)
		MoveLoc(2)
		Ret,
	public 0x0.LibraCoin.mint(Integer, &resource MintCapability@0x0.LibraCoin): (resource T@0x0.LibraCoin)
		locals(1): Integer, &resource MintCapability@0x0.LibraCoin,
		MoveLoc(1)
		ReleaseRef
		MoveLoc(0)
		Pack(0)
		Ret,
	0x0.LibraCoin.grant_mint_capability(): ()
		locals(2): ,
		Pack(1)
		MoveToSender(1)
		Ret,
	public 0x0.LibraCoin.zero(): (resource T@0x0.LibraCoin)
		locals(2): ,
		LdConst(0)
		Pack(0)
		Ret,
	public 0x0.LibraCoin.value(&resource T@0x0.LibraCoin): (Integer)
		locals(3): &resource T@0x0.LibraCoin,
		MoveLoc(0)
		BorrowField(resource T@0x0.LibraCoin.value: Integer)
		ReadRef
		Ret,
	public 0x0.LibraCoin.split(resource T@0x0.LibraCoin, Integer): (resource T@0x0.LibraCoin, resource T@0x0.LibraCoin)
		locals(4): resource T@0x0.LibraCoin, Integer, resource T@0x0.LibraCoin,
		BorrowLoc(0)
		MoveLoc(1)
		Call(0x0.LibraCoin.withdraw(&mut resource T@0x0.LibraCoin, Integer): (resource T@0x0.LibraCoin))
		StLoc(2)
		MoveLoc(0)
		MoveLoc(2)
		Ret,
	public 0x0.LibraCoin.withdraw(&mut resource T@0x0.LibraCoin, Integer): (resource T@0x0.LibraCoin)
		locals(5): &mut resource T@0x0.LibraCoin, Integer, Integer,
		CopyLoc(0)
		BorrowField(resource T@0x0.LibraCoin.value: Integer)
		ReadRef
		StLoc(2)
		LdConst(10)
		CopyLoc(2)
		CopyLoc(1)
		Ge
		Assert
		MoveLoc(2)
		CopyLoc(1)
		Sub
		MoveLoc(0)
		BorrowField(resource T@0x0.LibraCoin.value: Integer)
		WriteRef
		MoveLoc(1)
		Pack(0)
		Ret,
	public 0x0.LibraCoin.join(resource T@0x0.LibraCoin, resource T@0x0.LibraCoin): (resource T@0x0.LibraCoin)
		locals(6): resource T@0x0.LibraCoin, resource T@0x0.LibraCoin,
		BorrowLoc(0)
		MoveLoc(1)
		Call(0x0.LibraCoin.deposit(&mut resource T@0x0.LibraCoin, resource T@0x0.LibraCoin): ())
		MoveLoc(0)
		Ret,
	public 0x0.LibraCoin.deposit(&mut resource T@0x0.LibraCoin, resource T@0x0.LibraCoin): ()
		locals(7): &mut resource T@0x0.LibraCoin, resource T@0x0.LibraCoin, Integer, Integer,
		CopyLoc(0)
		BorrowField(resource T@0x0.LibraCoin.value: Integer)
		ReadRef
		StLoc(2)
		MoveLoc(1)
		Unpack(0)
		StLoc(3)
		MoveLoc(2)
		MoveLoc(3)
		Add
		MoveLoc(0)
		BorrowField(resource T@0x0.LibraCoin.value: Integer)
		WriteRef
		Ret,
	public 0x0.LibraCoin.destroy_zero(resource T@0x0.LibraCoin): ()
		locals(8): resource T@0x0.LibraCoin, Integer,
		MoveLoc(0)
		Unpack(0)
		StLoc(1)
		LdConst(11)
		MoveLoc(1)
		LdConst(0)
		Eq
		Assert
		Ret,
	public 0x0.LibraCoin.TODO_REMOVE_burn_gas_fee(resource T@0x0.LibraCoin): ()
		locals(8): resource T@0x0.LibraCoin, Integer,
		MoveLoc(0)
		Unpack(0)
		StLoc(1)
		Ret,]
Type Signatures: [
	Integer,]
Function Signatures: [
	(): (&resource MintCapability@0x0.LibraCoin),
	(Integer, &resource MintCapability@0x0.LibraCoin): (resource T@0x0.LibraCoin),
	(): (),
	(): (resource T@0x0.LibraCoin),
	(&resource T@0x0.LibraCoin): (Integer),
	(resource T@0x0.LibraCoin, Integer): (resource T@0x0.LibraCoin, resource T@0x0.LibraCoin),
	(&mut resource T@0x0.LibraCoin, Integer): (resource T@0x0.LibraCoin),
	(resource T@0x0.LibraCoin, resource T@0x0.LibraCoin): (resource T@0x0.LibraCoin),
	(&mut resource T@0x0.LibraCoin, resource T@0x0.LibraCoin): (),
	(resource T@0x0.LibraCoin): (),]
Locals Signatures: [
	Address, &mut resource MintCapability@0x0.LibraCoin, &resource MintCapability@0x0.LibraCoin,
	Integer, &resource MintCapability@0x0.LibraCoin,
	,
	&resource T@0x0.LibraCoin,
	resource T@0x0.LibraCoin, Integer, resource T@0x0.LibraCoin,
	&mut resource T@0x0.LibraCoin, Integer, Integer,
	resource T@0x0.LibraCoin, resource T@0x0.LibraCoin,
	&mut resource T@0x0.LibraCoin, resource T@0x0.LibraCoin, Integer, Integer,
	resource T@0x0.LibraCoin, Integer,]
Strings: [
	LibraCoin,
	T,
	value,
	MintCapability,
	borrow_sender_mint_capability,
	mint,
	grant_mint_capability,
	zero,
	split,
	withdraw,
	join,
	deposit,
	destroy_zero,
	TODO_REMOVE_burn_gas_fee,]
ByteArrays: []
Addresses: [
	0x0,]
}

