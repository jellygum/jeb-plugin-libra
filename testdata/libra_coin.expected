Module Handles: [
    0x0.LibraCoin,
]
Struct Handles: [
    resource T@0x0.LibraCoin,
    resource MintCapability@0x0.LibraCoin,
]
Function Handles: [
    0x0.LibraCoin.borrow_sender_mint_capability(): (& struct resource MintCapability@0x0.LibraCoin),
    0x0.LibraCoin.mint(integer, & struct resource MintCapability@0x0.LibraCoin): (struct resource T@0x0.LibraCoin),
    0x0.LibraCoin.grant_mint_capability(): (),
    0x0.LibraCoin.zero(): (struct resource T@0x0.LibraCoin),
    0x0.LibraCoin.value(& struct resource T@0x0.LibraCoin): (integer),
    0x0.LibraCoin.split(struct resource T@0x0.LibraCoin, integer): (struct resource T@0x0.LibraCoin, struct resource T@0x0.LibraCoin),
    0x0.LibraCoin.withdraw(&mut struct resource T@0x0.LibraCoin, integer): (struct resource T@0x0.LibraCoin),
    0x0.LibraCoin.join(struct resource T@0x0.LibraCoin, struct resource T@0x0.LibraCoin): (struct resource T@0x0.LibraCoin),
    0x0.LibraCoin.deposit(&mut struct resource T@0x0.LibraCoin, struct resource T@0x0.LibraCoin): (),
    0x0.LibraCoin.destroy_zero(struct resource T@0x0.LibraCoin): (),
    0x0.LibraCoin.TODO_REMOVE_burn_gas_fee(struct resource T@0x0.LibraCoin): (),
]
Struct Definitions: [
    resource T@0x0.LibraCoin { value: integer },
    resource MintCapability@0x0.LibraCoin {  },
]
Field Definitions: [
    value: integer,
]
Type Signatures: [
    integer,
]
Function Signatures: [
    (): (& struct resource MintCapability@0x0.LibraCoin),
    (integer, & struct resource MintCapability@0x0.LibraCoin): (struct resource T@0x0.LibraCoin),
    (): (),
    (): (struct resource T@0x0.LibraCoin),
    (& struct resource T@0x0.LibraCoin): (integer),
    (struct resource T@0x0.LibraCoin, integer): (struct resource T@0x0.LibraCoin, struct resource T@0x0.LibraCoin),
    (&mut struct resource T@0x0.LibraCoin, integer): (struct resource T@0x0.LibraCoin),
    (struct resource T@0x0.LibraCoin, struct resource T@0x0.LibraCoin): (struct resource T@0x0.LibraCoin),
    (&mut struct resource T@0x0.LibraCoin, struct resource T@0x0.LibraCoin): (),
    (struct resource T@0x0.LibraCoin): (),
]
Local Signatures: [
    address, &mut struct resource MintCapability@0x0.LibraCoin, & struct resource MintCapability@0x0.LibraCoin,
    integer, & struct resource MintCapability@0x0.LibraCoin,
    ,
    & struct resource T@0x0.LibraCoin,
    struct resource T@0x0.LibraCoin, integer, struct resource T@0x0.LibraCoin,
    &mut struct resource T@0x0.LibraCoin, integer, integer,
    struct resource T@0x0.LibraCoin, struct resource T@0x0.LibraCoin,
    &mut struct resource T@0x0.LibraCoin, struct resource T@0x0.LibraCoin, integer, integer,
    struct resource T@0x0.LibraCoin, integer,
]
Strings: [
    LibraCoin,
    T,
    value,
    MintCapability,
    borrow_sender_mint_capability,
    mint,
    grant_mint_capability,
    zero,
    split,
    withdraw,
    join,
    deposit,
    destroy_zero,
    TODO_REMOVE_burn_gas_fee,
]
ByteArrays: [
]
Addresses: [
    0x0,
]
Function Definitions: [
    public 0x0.LibraCoin.borrow_sender_mint_capability(): (& struct resource MintCapability@0x0.LibraCoin)
        max_stack_size: 4
        locals: address, &mut struct resource MintCapability@0x0.LibraCoin, & struct resource MintCapability@0x0.LibraCoin
        GET_TXN_SENDER                     [0,1]
        ST_LOC IdxLocal(0)                 [1,0]
        MOVE_LOC IdxLocal(0)               [0,1]
        BORROW_REF IdxStructDef(1)         [1,1]
        ST_LOC IdxLocal(1)                 [1,0]
        MOVE_LOC IdxLocal(1)               [0,1]
        FREEZE_REF                         [1,1]
        ST_LOC IdxLocal(2)                 [1,0]
        MOVE_LOC IdxLocal(2)               [0,1]
        RET                                [1,0]
    ,
    public 0x0.LibraCoin.mint(integer, & struct resource MintCapability@0x0.LibraCoin): (struct resource T@0x0.LibraCoin)
        max_stack_size: 2
        locals: integer, & struct resource MintCapability@0x0.LibraCoin
        MOVE_LOC IdxLocal(1)               [0,1]
        RELEASE_REF                        [1,0]
        MOVE_LOC IdxLocal(0)               [0,1]
        PACK IdxStructDef(0)               [1,1]
        RET                                [1,0]
    ,
    0x0.LibraCoin.grant_mint_capability(): ()
        max_stack_size: 2
        locals: 
        PACK IdxStructDef(1)               [0,1]
        MOVE_TO IdxStructDef(1)            [1,0]
        RET                                [0,0]
    ,
    public 0x0.LibraCoin.zero(): (struct resource T@0x0.LibraCoin)
        max_stack_size: 1
        locals: 
        LD_CONST ImmUint64(0)              [0,1]
        PACK IdxStructDef(0)               [1,1]
        RET                                [1,0]
    ,
    public 0x0.LibraCoin.value(& struct resource T@0x0.LibraCoin): (integer)
        max_stack_size: 1
        locals: & struct resource T@0x0.LibraCoin
        MOVE_LOC IdxLocal(0)               [0,1]
        LD_REF_FIELD IdxFieldDef(0)        [1,1]
        READ_REF                           [1,1]
        RET                                [1,0]
    ,
    public 0x0.LibraCoin.split(struct resource T@0x0.LibraCoin, integer): (struct resource T@0x0.LibraCoin, struct resource T@0x0.LibraCoin)
        max_stack_size: 3
        locals: struct resource T@0x0.LibraCoin, integer, struct resource T@0x0.LibraCoin
        LD_REF_LOC IdxLocal(0)             [0,1]
        MOVE_LOC IdxLocal(1)               [1,2]
        CALL IdxFuncHandle(6)              [2,1]
        ST_LOC IdxLocal(2)                 [1,0]
        MOVE_LOC IdxLocal(0)               [0,1]
        MOVE_LOC IdxLocal(2)               [1,2]
        RET                                [2,0]
    ,
    public 0x0.LibraCoin.withdraw(&mut struct resource T@0x0.LibraCoin, integer): (struct resource T@0x0.LibraCoin)
        max_stack_size: 3
        locals: &mut struct resource T@0x0.LibraCoin, integer, integer
        COPY_LOC IdxLocal(0)               [0,1]
        LD_REF_FIELD IdxFieldDef(0)        [1,1]
        READ_REF                           [1,1]
        ST_LOC IdxLocal(2)                 [1,0]
        LD_CONST ImmUint64(10)             [0,1]
        COPY_LOC IdxLocal(2)               [1,2]
        COPY_LOC IdxLocal(1)               [2,3]
        GE                                 [3,2]
        ASSERT                             [2,0]
        MOVE_LOC IdxLocal(2)               [0,1]
        COPY_LOC IdxLocal(1)               [1,2]
        SUB                                [2,1]
        MOVE_LOC IdxLocal(0)               [1,2]
        LD_REF_FIELD IdxFieldDef(0)        [2,2]
        WRITE_REF                          [2,0]
        MOVE_LOC IdxLocal(1)               [0,1]
        PACK IdxStructDef(0)               [1,1]
        RET                                [1,0]
    ,
    public 0x0.LibraCoin.join(struct resource T@0x0.LibraCoin, struct resource T@0x0.LibraCoin): (struct resource T@0x0.LibraCoin)
        max_stack_size: 2
        locals: struct resource T@0x0.LibraCoin, struct resource T@0x0.LibraCoin
        LD_REF_LOC IdxLocal(0)             [0,1]
        MOVE_LOC IdxLocal(1)               [1,2]
        CALL IdxFuncHandle(8)              [2,0]
        MOVE_LOC IdxLocal(0)               [0,1]
        RET                                [1,0]
    ,
    public 0x0.LibraCoin.deposit(&mut struct resource T@0x0.LibraCoin, struct resource T@0x0.LibraCoin): ()
        max_stack_size: 2
        locals: &mut struct resource T@0x0.LibraCoin, struct resource T@0x0.LibraCoin, integer, integer
        COPY_LOC IdxLocal(0)               [0,1]
        LD_REF_FIELD IdxFieldDef(0)        [1,1]
        READ_REF                           [1,1]
        ST_LOC IdxLocal(2)                 [1,0]
        MOVE_LOC IdxLocal(1)               [0,1]
        UNPACK IdxStructDef(0)             [1,1]
        ST_LOC IdxLocal(3)                 [1,0]
        MOVE_LOC IdxLocal(2)               [0,1]
        MOVE_LOC IdxLocal(3)               [1,2]
        ADD                                [2,1]
        MOVE_LOC IdxLocal(0)               [1,2]
        LD_REF_FIELD IdxFieldDef(0)        [2,2]
        WRITE_REF                          [2,0]
        RET                                [0,0]
    ,
    public 0x0.LibraCoin.destroy_zero(struct resource T@0x0.LibraCoin): ()
        max_stack_size: 3
        locals: struct resource T@0x0.LibraCoin, integer
        MOVE_LOC IdxLocal(0)               [0,1]
        UNPACK IdxStructDef(0)             [1,1]
        ST_LOC IdxLocal(1)                 [1,0]
        LD_CONST ImmUint64(11)             [0,1]
        MOVE_LOC IdxLocal(1)               [1,2]
        LD_CONST ImmUint64(0)              [2,3]
        EQ                                 [3,2]
        ASSERT                             [2,0]
        RET                                [0,0]
    ,
    public 0x0.LibraCoin.TODO_REMOVE_burn_gas_fee(struct resource T@0x0.LibraCoin): ()
        max_stack_size: 1
        locals: struct resource T@0x0.LibraCoin, integer
        MOVE_LOC IdxLocal(0)               [0,1]
        UNPACK IdxStructDef(0)             [1,1]
        ST_LOC IdxLocal(1)                 [1,0]
        RET                                [0,0]
    ,
]

